laurent22/joplin
# Desktop: Fixes #8916 Added 'save to file' after keyboard shortcuts are imported

## Overview
Keyboard shortcuts that are imported, will now save to file after being imported.

Previously after importing keyboard shortcuts, the shortcuts would **not** be saved to file. This forced users to re-import their keyboard shortcuts each time Joplin was restarted
- As a workaround users could trigger the save by importing shortcuts, then modifying a single shortcut in the GUI to trigger persistence.

## What changed and why?
Reviewing the [code comments](https://github.com/laurent22/joplin/blob/0638922ae44d999eb7c3f5fd5faa0f3fa5b9c96f/packages/app-desktop/gui/KeymapConfig/utils/useKeymap.ts#L63), it appears that 'save to file after import' was originally intended, however [`setKeymapItems(keymapService.getKeymapItems());`](https://github.com/laurent22/joplin/blob/0638922ae44d999eb7c3f5fd5faa0f3fa5b9c96f/packages/app-desktop/gui/KeymapConfig/utils/useKeymap.ts#L64) is not enough to trigger to save to file, as [saves can only be triggered with `setMustSave(true);`](https://github.com/laurent22/joplin/blob/0638922ae44d999eb7c3f5fd5faa0f3fa5b9c96f/packages/app-desktop/gui/KeymapConfig/utils/useKeymap.ts#L74)


This change brings [overrideKeymapItems()](https://github.com/laurent22/joplin/blob/0638922ae44d999eb7c3f5fd5faa0f3fa5b9c96f/packages/app-desktop/gui/KeymapConfig/utils/useKeymap.ts#L54) in line with [resetAccelerator()](https://github.com/laurent22/joplin/blob/0638922ae44d999eb7c3f5fd5faa0f3fa5b9c96f/packages/app-desktop/gui/KeymapConfig/utils/useKeymap.ts#L42C8-L42C24) and [setAccelerator()](https://github.com/laurent22/joplin/blob/0638922ae44d999eb7c3f5fd5faa0f3fa5b9c96f/packages/app-desktop/gui/KeymapConfig/utils/useKeymap.ts#L31) which correctly call `setMustSave(true);` after `setKeymapItems()`

## Test Steps
1. Navigate to Tools > Options > Keyboard Shortcuts
2. Verify that the `Statistics...` shortcut is `Disabled`
3. Select `Import`
4. Import test file [TEST-Shortcuts.json](https://github.com/laurent22/joplin/files/13160769/TEST-Shortcuts.json)
5. Verify that the `Statistics...` command is assigned to `F5`
6. Restart Joplin
7. Navigate to Tools > Options > Keyboard Shortcuts
8. Verify that command `Statistics...` is still assigned to `F5`
9. Try the F5 shortcut to open statistics menu



Fixes #8916
